File: frontend/src/api/mastery.ts
import { apiClient } from "./client";

export interface MasteryItem {
  conceptId: string;
  conceptName: string;
  masteryScore: number;
  confidence: number;
  trend: "IMPROVING" | "STABLE" | "DECLINING";
  evidenceCount: number;
  recentPerformance: string[];
  hasEvidence: boolean;
  updatedAt: string | null;
}

export interface MasteryDetail extends MasteryItem {
  evidenceQuestions: number;
  contributingAssessments: string[];
}

export interface MasteryHistoryItem {
  id: string;
  assessmentId: string | null;
  previousScore: number | null;
  newScore: number;
  confidence: number;
  createdAt: string;
}

interface BackendMastery {
  concept_id: string;
  concept_name: string;
  mastery_score: number;
  confidence: number;
  trend: MasteryItem["trend"];
  evidence_count: number;
  recent_performance: string[];
  has_evidence: boolean;
  updated_at: string | null;
  evidence_questions?: number;
  contributing_assessments?: string[];
}

interface BackendHistory {
  id: string;
  assessment_id: string | null;
  previous_score: number | null;
  new_score: number;
  confidence: number;
  created_at: string;
}

interface Page<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

function toItem(m: BackendMastery): MasteryItem {
  return {
    conceptId: m.concept_id,
    conceptName: m.concept_name,
    masteryScore: m.mastery_score,
    confidence: m.confidence,
    trend: m.trend,
    evidenceCount: m.evidence_count,
    recentPerformance: m.recent_performance ?? [],
    hasEvidence: m.has_evidence,
    updatedAt: m.updated_at,
  };
}

export const masteryApi = {
  async list(
    projectId: string,
    params: { page?: number; pageSize?: number; sort?: string } = {},
  ): Promise<{ items: MasteryItem[]; total: number }> {
    const res = await apiClient.get<Page<BackendMastery>>(`/api/v1/projects/${projectId}/mastery`, {
      params: {
        page: params.page ?? 1,
        page_size: params.pageSize ?? 20,
        sort: params.sort ?? "lowest",
      },
    });
    return { items: res.data.items.map(toItem), total: res.data.total };
  },

  async detail(projectId: string, conceptId: string): Promise<MasteryDetail> {
    const res = await apiClient.get<BackendMastery>(
      `/api/v1/projects/${projectId}/mastery/${conceptId}`,
    );
    return {
      ...toItem(res.data),
      evidenceQuestions: res.data.evidence_questions ?? 0,
      contributingAssessments: res.data.contributing_assessments ?? [],
    };
  },

  async history(projectId: string, conceptId: string): Promise<MasteryHistoryItem[]> {
    const res = await apiClient.get<Page<BackendHistory>>(
      `/api/v1/projects/${projectId}/mastery/${conceptId}/history`,
      { params: { page_size: 50 } },
    );
    return res.data.items.map((h) => ({
      id: h.id,
      assessmentId: h.assessment_id,
      previousScore: h.previous_score,
      newScore: h.new_score,
      confidence: h.confidence,
      createdAt: h.created_at,
    }));
  },
};

File: backend/app/api/routes/mastery.py
"""Mastery endpoints: per-concept estimates, detail explanations, history.

Every route verifies authenticated user -> owned project -> in-project
concept. Cross-tenant access returns 404 without disclosure. All numbers are
computed server-side by MasteryService (deterministic, no LLM); there is no
endpoint that writes mastery directly — estimates only move through
assessment completion.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.auth.dependencies import CurrentUser, get_owned_project
from app.db.session import get_db
from app.models.learning import Project
from app.schemas.common import Page
from app.schemas.mastery import MasteryDetailRead, MasteryHistoryRead, MasteryRead
from app.services.mastery_service import MasteryService

router = APIRouter(prefix="/projects", tags=["mastery"])

SessionDep = Annotated[Session, Depends(get_db)]
OwnedProject = Annotated[Project, Depends(get_owned_project)]


def _to_read(item: dict) -> MasteryRead:
    mastery = item["mastery"]
    return MasteryRead(
        concept_id=mastery.concept_id,
        concept_name=item["concept_name"],
        mastery_score=round(min(max(mastery.score / 100, 0.0), 1.0), 4),
        confidence=round(min(max(mastery.confidence, 0.0), 1.0), 4),
        trend=mastery.trend,
        evidence_count=item["evidence_count"],
        recent_performance=[],
        has_evidence=True,
        updated_at=mastery.updated_at,
    )


@router.get("/{project_id}/mastery", response_model=Page[MasteryRead])
def list_mastery(
    project: OwnedProject,
    user: CurrentUser,
    session: SessionDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    sort: Annotated[str, Query(pattern="^(lowest|recent|name)$")] = "lowest",
):
    service = MasteryService(session)
    items, total = service.list_mastery(
        user_id=user.id, project_id=project.id, page=page, page_size=page_size, sort=sort
    )
    return Page[MasteryRead](
        items=[_to_read(item) for item in items], total=total, page=page, page_size=page_size
    )


@router.get("/{project_id}/mastery/{concept_id}", response_model=MasteryDetailRead)
def get_mastery(
    concept_id: uuid.UUID, project: OwnedProject, user: CurrentUser, session: SessionDep
):
    service = MasteryService(session)
    explanation = service.explain(user_id=user.id, project_id=project.id, concept_id=concept_id)
    return MasteryDetailRead(
        concept_id=explanation.concept_id,
        concept_name=explanation.concept_name,
        mastery_score=explanation.mastery_score,
        confidence=explanation.confidence,
        trend=explanation.trend,
        evidence_count=explanation.evidence_assessments,
        recent_performance=list(explanation.recent_performance),
        has_evidence=explanation.has_evidence,
        updated_at=explanation.updated_at,
        evidence_questions=explanation.evidence_questions,
        contributing_assessments=list(explanation.contributing_assessments),
    )


@router.get(
    "/{project_id}/mastery/{concept_id}/history",
    response_model=Page[MasteryHistoryRead],
)
def get_mastery_history(
    concept_id: uuid.UUID,
    project: OwnedProject,
    user: CurrentUser,
    session: SessionDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
):
    service = MasteryService(session)
    rows, total = service.mastery_history(
        user_id=user.id,
        project_id=project.id,
        concept_id=concept_id,
        page=page,
        page_size=page_size,
    )
    return Page[MasteryHistoryRead](
        items=[
            MasteryHistoryRead(
                id=row.id,
                assessment_id=row.assessment_id,
                previous_score=(
                    round(row.previous_score / 100, 4) if row.previous_score is not None else None
                ),
                new_score=round(min(max(row.score / 100, 0.0), 1.0), 4),
                confidence=round(min(max(row.confidence, 0.0), 1.0), 4),
                created_at=row.created_at,
            )
            for row in rows
        ],
        total=total,
        page=page,
        page_size=page_size,
    )
